overlord/devicestate/devicemgr.go

```diff
@@ -1804,48 +1804,70 @@ func (m *DeviceManager) Systems() ([]*System, error) {
+		return nil, nil, nil, fmt.Errorf("cannot get model and gadget information on a classic boot system")
 	}
 	// get current system as input for loadSeedAndSystem()
 	systemMode := m.SystemMode(SysAny)
 	currentSys, _ := currentSystemForMode(m.state, systemMode)
 
 	s, sys, err := loadSeedAndSystem(wantedSystemLabel, currentSys)
 	if err != nil {
-		return nil, nil, err
+		return nil, nil, nil, err
 	}
 
 	// 2. get the gadget volumes for the given seed-label
 	perf := &timings.Timings{}
 	if err := s.LoadEssentialMeta([]snap.Type{snap.TypeGadget}, perf); err != nil {
-		return nil, nil, fmt.Errorf("cannot load gadget snap metadata: %v", err)
+		return nil, nil, nil, fmt.Errorf("cannot load gadget snap metadata: %v", err)
 	}
 	gadgetSnap := s.EssentialSnaps()[0]
 	if gadgetSnap.Path == "" {
-		return nil, nil, fmt.Errorf("internal error: cannot get gadget snap path")
+		return nil, nil, nil, fmt.Errorf("internal error: cannot get gadget snap path")
 	}
 	snapf, err := snapfile.Open(gadgetSnap.Path)
 	if err != nil {
-		return nil, nil, fmt.Errorf("cannot open gadget snap: %v", err)
+		return nil, nil, nil, fmt.Errorf("cannot open gadget snap: %v", err)
 	}
 	gadgetYaml, err := snapf.ReadFile("meta/gadget.yaml")
 	if err != nil {
-		return nil, nil, fmt.Errorf("cannot read gadget.yaml: %v", err)
+		return nil, nil, nil, fmt.Errorf("cannot read gadget.yaml: %v", err)
 	}
 	gadgetInfo, err := gadget.InfoFromGadgetYaml(gadgetYaml, sys.Model)
 	if err != nil {
-		return nil, nil, fmt.Errorf("cannot parse gadget.yaml: %v", err)
+		return nil, nil, nil, fmt.Errorf("cannot parse gadget.yaml: %v", err)
+	}
+
+	// need kernel details to validate storage against encryption
+	if err := s.LoadEssentialMeta([]snap.Type{snap.TypeKernel}, perf); err != nil {
+		return nil, nil, nil, fmt.Errorf("cannot load kernel snap metadata: %v", err)
+	}
+	kernelSnap := s.EssentialSnaps()[0]
+	if kernelSnap.Path == "" {
+		return nil, nil, nil, fmt.Errorf("internal error: cannot get kernel snap path")
+	}
+	snapf, err = snapfile.Open(gadgetSnap.Path)
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf("cannot open gadget snap: %v", err)
+	}
+	kernelInfo, err := snap.ReadInfoFromSnapFile(snapf, kernelSnap.SideInfo)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	encInfo, err := m.EncryptionSupportInfo(sys.Model, kernelInfo, gadgetInfo)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	opts := &gadget.ValidationConstraints{
+		EncryptedData: encInfo.StorageSafety == asserts.StorageSafetyEncrypted,
 	}
-	// TODO: once EncryptionSupportInfo from PR#12060 is available use it
-	//       here to set the right option
-	opts := &gadget.ValidationConstraints{}
 	if err := gadget.Validate(gadgetInfo, sys.Model, opts); err != nil {
-		return nil, nil, fmt.Errorf("cannot validate gadget.yaml: %v", err)
+		return nil, nil, nil, fmt.Errorf("cannot validate gadget.yaml: %v", err)
 	}
 
-	return sys, gadgetInfo, nil
+	return sys, gadgetInfo, kernelInfo, nil
```

Comment:
shouldn't this be made to return directly the encInfo instead of the KernelInfo to ask again in daemon?

Reply:
Indeed, much simpler this way, thank you